compiler/rustc_passes/src/dead.rs

```diff
@@ -487,32 +489,63 @@ impl<'tcx> MarkSymbolVisitor<'tcx> {
 
     fn solve_rest_impl_items(&mut self, mut unsolved_impl_items: Vec<(hir::ItemId, LocalDefId)>) {
         let mut ready;
-        (ready, unsolved_impl_items) = unsolved_impl_items
-            .into_iter()
-            .partition(|&(impl_id, _)| self.impl_item_with_used_self(impl_id));
+        (ready, unsolved_impl_items) =
+            unsolved_impl_items.into_iter().partition(|&(impl_id, impl_item_id)| {
+                self.impl_item_with_used_self_and_trait_item(impl_id, impl_item_id)
+            });
 
         while !ready.is_empty() {
             self.worklist =
                 ready.into_iter().map(|(_, id)| (id, ComesFromAllowExpect::No)).collect();
             self.mark_live_symbols();
 
-            (ready, unsolved_impl_items) = unsolved_impl_items
-                .into_iter()
-                .partition(|&(impl_id, _)| self.impl_item_with_used_self(impl_id));
+            (ready, unsolved_impl_items) =
+                unsolved_impl_items.into_iter().partition(|&(impl_id, impl_item_id)| {
+                    self.impl_item_with_used_self_and_trait_item(impl_id, impl_item_id)
+                });
         }
     }
 
-    fn impl_item_with_used_self(&mut self, impl_id: hir::ItemId) -> bool {
+    fn impl_item_with_used_self_and_trait_item(
+        &mut self,
+        impl_id: hir::ItemId,
+        impl_item_id: LocalDefId,
+    ) -> bool {
         if let TyKind::Path(hir::QPath::Resolved(_, path)) =
             self.tcx.hir().item(impl_id).expect_impl().self_ty.kind
             && let Res::Def(def_kind, def_id) = path.res
             && let Some(local_def_id) = def_id.as_local()
             && matches!(def_kind, DefKind::Struct | DefKind::Enum | DefKind::Union)
         {
-            self.live_symbols.contains(&local_def_id)
-        } else {
-            false
+            if self.tcx.visibility(impl_item_id).is_public() {
+                // for the public method, we don't know the trait item is used or not,
+                // so we mark the method live if the self is used
+                return self.live_symbols.contains(&local_def_id);
+            }
+
+            let impl_id = impl_id.owner_id.to_def_id();
+            let impl_item_trait_item_ids = self
+                .impl_item_trait_item_ids_cache
+                .entry(impl_id)
+                .or_insert_with(|| Self::impl_item_trait_item_ids(self.tcx, impl_id));
+
+            if let Some(trait_item_id) = impl_item_trait_item_ids.get(&impl_item_id.to_def_id())
```

Comment:
```suggestion
            if let Some(trait_item_id) = self.tcx.associated_item(impl_item_id).trait_item_def_id
```

Reply:
Great! I think the description of this method is a bit confused, I skipped it cause it said "Maps from a trait item to the trait item “descriptor”", I thought it only works for trait items.

Reply:
Maybe this doc can be updated?

Reply:
Yes, if the doc is unclear then it's better to updated it.